File: increasingring/src/main/java/com/baybaka/increasingring/config/CachingConfigFactory.java
package com.baybaka.increasingring.config;

import android.media.AudioManager;

import com.baybaka.increasingring.SettingsService;

import javax.inject.Inject;

public class CachingConfigFactory {

    private SettingsService settingsService;

    private final RingerConfig config;
    private int maxHardwareVolumeLevel;
    private int oldPreRingLevel;

    private boolean isMinVolLimited;
    private boolean isMinLimitToPreRing;
    private int minVolLimit;

    private boolean isMaxVolLimited;
    private boolean isMaxLimitToPrering;
    private int maxVolLimit;


    @Inject
    public CachingConfigFactory(SettingsService settingsService) {
        this.settingsService = settingsService;
        config = new RingerConfig();
    }

    public void setMaxHardwareVolumeLevel(int maxHardwareVolumeLevel) {
        this.maxHardwareVolumeLevel = maxHardwareVolumeLevel;
    }

    public RingerConfig getConfig(int preRingLevel) {

        if (preRingLevel != oldPreRingLevel) {
            if (isMinLimitToPreRing || isMaxLimitToPrering) {
                updateMinMax(preRingLevel);
            }
            oldPreRingLevel = preRingLevel;
        }
        return config;
    }

    public RingerConfig getFindPhoneConfig(){
        RingerConfig config = new RingerConfig();
        config.setStartSoundLevel(14);
        config.setAllowedMaxVolume(14);
        config.setUseMusicStream(true);
        config.setInterval(1);
        config.setVibrateFirst(false);
        config.setMuteFirst(false);
        return config;
    }


    public void updateConfig() {

        localConfigUpdate();

        config.setInterval(readIntervalConfig());

        config.setMuteFirst(readMuteConfig());
        config.setMuteTimes(readMuteTimes());

        config.setVibrateFirst(readVibrateConfig());
        config.setVibrateTimes(readVibrateTimes());
        // in case of change upper volume limit
        updateMinMax(1);
    }

    private int readMuteTimes() {
        return settingsService.getMuteTimesCount();
    }

    private int readVibrateTimes() {
        return settingsService.getVibrateTimesCount();
    }

    private void localConfigUpdate(){
        minVolLimit = settingsService.getMinVolumeLimit();
        isMinVolLimited = settingsService.isMinVolumeLimited();

        isMinLimitToPreRing = isMinVolLimited && isLimitEqualsToPreRing(minVolLimit);

        maxVolLimit = settingsService.getMaxVolumeLimit();
        isMaxVolLimited = settingsService.isMaxVolumeLimited();

        isMaxLimitToPrering = isMaxVolLimited && isLimitEqualsToPreRing(maxVolLimit);

    }

    private void updateMinMax(int preRingLevel) {

        int minLimitCalc = 1;
        if(isMinVolLimited) {
            minLimitCalc = isMinLimitToPreRing ? preRingLevel : minVolLimit;
        }

        int maxLimitCalc = maxHardwareVolumeLevel;
        if(isMaxVolLimited) {
            maxLimitCalc = isMaxLimitToPrering ? preRingLevel : maxVolLimit;
        }

        config.setStartSoundLevel(minLimitCalc);
        config.setAllowedMaxVolume(maxLimitCalc);
    }





    private boolean isLimitEqualsToPreRing(int resultLevel) {
        return resultLevel == maxHardwareVolumeLevel + 1;
    }

    private int readIntervalConfig() {
        return settingsService.getInterval();
    }

    private boolean readMuteConfig() {
        return settingsService.isMuteFirst();
    }

    private boolean readVibrateConfig() {
        return settingsService.isVibrateFirst();
    }

    public void updateStream(int soundStream) {
        if (soundStream == AudioManager.STREAM_MUSIC) {
            config.setUseMusicStream(true);
        } else {
            config.setUseMusicStream(false);
        }
    }


}

File: increasingring/src/main/java/com/baybaka/increasingring/config/CachingConfigFactory.kt
package com.baybaka.increasingring.config

import android.media.AudioManager
import com.baybaka.increasingring.audio.IAudioController
import com.baybaka.increasingring.settings.SettingsService
import org.slf4j.LoggerFactory
import javax.inject.Inject

class CachingConfigFactory @Inject
constructor(private val settingsService: SettingsService,
            private val audio: IAudioController) {

    private val config: RingerConfig = RingerConfig()
    private var maxHardwareVolumeLevel: Int = 0
    private var oldPreRingLevel: Int = 0

    private var isMinVolLimited: Boolean = false
    private var isMinLimitToPreRing: Boolean = false
    private var minVolLimit: Int = 1

    private var isMaxVolLimited: Boolean = false
    private var isMaxLimitToPreRing: Boolean = false
    private var maxVolLimit: Int = 99


    private val LOG = LoggerFactory.getLogger(CachingConfigFactory::class.java.simpleName)


    fun updateMaxHardwareVolumeLevel() {
        this.maxHardwareVolumeLevel = audio.new_GetMaxLevel()
    }


    internal fun currentConfig(): RingerConfig {
        val preRingLevel = getCurrentChosenStreamVolume()

        if (preRingLevel != oldPreRingLevel) {
            if (isMinLimitToPreRing || isMaxLimitToPreRing) {
                updateMinMaxForConfig(preRingLevel)
            }
            oldPreRingLevel = preRingLevel
        }
        return config
    }

    fun getConfig(): RingerConfig {
        return config
    }


    internal val findPhoneConfig: RingerConfig
        get() = RingerConfig().apply {
            startSoundLevel = 14
            allowedMaxVolume = 14
            useMusicStream = true && settingsService.canUseMusicStream()
            interval = 1
            isVibrateFirst = false
            isMuteFirst = false
        }



    fun updateConfig() {
        //        mAudioManagerWrapper.changeOutputStream(soundStream)
        //        configFactory.updateStream(mSettingsService.soundStream)
        updateStream(settingsService.soundStream)

        //        configFactory.setMaxHardwareVolumeLevel(mAudioManagerWrapper.chosenStreamrMaxHardwareVolumeLevel)
        updateMaxHardwareVolumeLevel()

        localConfigUpdate()

        config.interval = settingsService.interval

        config.isMuteFirst = settingsService.isMuteFirst
        config.muteTimes = settingsService.muteTimesCount

        config.isVibrateFirst = settingsService.isVibrateFirst
        config.vibrateTimes = settingsService.vibrateTimesCount

        config.doNotRing = settingsService.isSkipRing

        // in case of change upper volume limit
        updateMinMaxForConfig(1)
    }


    private fun localConfigUpdate() {
        minVolLimit = settingsService.minVolumeLimit
        isMinVolLimited = settingsService.isMinVolumeLimited

        isMinLimitToPreRing = isMinVolLimited && isLimitEqualsToPreRing(minVolLimit)

        maxVolLimit = settingsService.maxVolumeLimit
        isMaxVolLimited = settingsService.isMaxVolumeLimited

        isMaxLimitToPreRing = isMaxVolLimited && isLimitEqualsToPreRing(maxVolLimit)

    }

    private fun updateMinMaxForConfig(preRingLevel: Int) {

        config.startSoundLevel = when {
            isMinVolLimited && isMinLimitToPreRing -> preRingLevel
            isMaxVolLimited -> minVolLimit
            else -> 1
        }

        config.allowedMaxVolume = when{
            isMaxVolLimited && isMaxLimitToPreRing -> preRingLevel
            isMaxVolLimited -> maxVolLimit
            else -> maxHardwareVolumeLevel
        }
    }


    private fun isLimitEqualsToPreRing(resultLevel: Int): Boolean {
        return resultLevel == maxHardwareVolumeLevel + 1
    }


    private fun updateStream(soundStream: Int) {
        config.useMusicStream = soundStream == AudioManager.STREAM_MUSIC
    }

    private fun getCurrentChosenStreamVolume(): Int = audio.new_GetAudioLevel()


    fun printDebug() {
        LOG.debug("Current volume is $config")
        LOG.debug("Config is ${getCurrentChosenStreamVolume()}")
    }
}
